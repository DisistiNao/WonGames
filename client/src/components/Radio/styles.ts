import styled, { css } from 'styled-components'
import { RadioProps } from '.'

export const Wrapper = styled.main`
  display: flex;
  align-items: center;
`

export const Input = styled.input`
  ${({ theme }) => css`
    cursor: pointer;
    display: flex;
    align-items: center;
    justify-content: center;
    appearance: none;
    width: 1.8rem;
    height: 1.8rem;
    border: 0.2rem solid ${theme.colors.darkGray};
    border-radius: 50%;
    outline: none;

    &:focus {
      box-shadow: 0 0 0.5rem ${theme.colors.primary};
    }

    &::after {
      background: ${theme.colors.primary};
      content: '';
      position: absolute;

      width: 1rem;
      height: 1rem;
      border-radius: 50%;
      transition: opacity ${theme.transition.fast};
      opacity: 0;
    }

    &:checked {
      border-color: ${theme.colors.primary};
      position: relative;

      &::after {
        opacity: 1;
      }
    }
  `}
`

export const Label = styled.label<Pick<RadioProps, 'labelColors'>>`
  ${({ theme, labelColors }) => css`
    cursor: pointer;
    padding-left: ${theme.spacings.xxsmall};
    color: ${theme.colors[labelColors!]};
    line-height: 1.8rem;
  `}
`
