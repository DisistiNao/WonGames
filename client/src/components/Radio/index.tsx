import { InputHTMLAttributes } from 'react'

import * as S from './styles'

type RadioValue = string | ReadonlyArray<string> | number

export type RadioProps = {
  onCheck?: (value?: RadioValue) => void
  label?: string
  labelFor?: string
  labelColors?: 'black' | 'white'
  value?: RadioValue
} & InputHTMLAttributes<HTMLInputElement>

const Radio = ({
  onCheck,
  label,
  labelFor = '',
  labelColors = 'white',
  value,
  ...props
}: RadioProps) => {
  const onChange = () => {
    !!onCheck && onCheck(value)
  }

  return (
    <S.Wrapper>
      <S.Input
        type="radio"
        value={value}
        id={labelFor}
        onChange={onChange}
        {...props}
      />
      {!!label && (
        <S.Label htmlFor={labelFor} labelColors={labelColors}>
          {label}
        </S.Label>
      )}
    </S.Wrapper>
  )
}

export default Radio
