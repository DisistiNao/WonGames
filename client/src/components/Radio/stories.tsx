import { Story, Meta } from '@storybook/react'
import Radio, { RadioProps } from '.'

export default {
  title: 'Radio',
  component: Radio,
  parameters: {
    layout: 'fullscreen',
    backgrounds: {
      default: 'won-dark'
    }
  },
  argTypes: {
    onCheck: { action: 'checked' }
  }
} as Meta

export const Default: Story<RadioProps> = (args) => (
  <>
    <div style={{ padding: 10 }}>
      <Radio
        defaultChecked
        label="primeiro"
        labelFor="primeiro"
        id="primeiro"
        value="primeiro"
        name="name"
        {...args}
      />
    </div>
    <div style={{ padding: 10 }}>
      <Radio
        label="segundo"
        labelFor="segundo"
        id="segundo"
        value="segundo"
        name="name"
        {...args}
      />
    </div>
    <div style={{ padding: 10 }}>
      <Radio
        label="terceiro"
        labelFor="terceiro"
        id="terceiro"
        value="terceiro"
        name="name"
        {...args}
      />
    </div>
  </>
)
