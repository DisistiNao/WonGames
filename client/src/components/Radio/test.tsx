import { screen, waitFor } from '@testing-library/react'
import { renderWithTheme } from 'utils/tests/helpers'
import theme from 'styles/theme'

import Radio from '.'
import userEvent from '@testing-library/user-event'

describe('<Radio />', () => {
  it('should render with label', () => {
    const { container } = renderWithTheme(
      <Radio label="Radio label" labelFor="radio" />
    )

    expect(screen.getByRole('radio')).toBeInTheDocument()
    expect(screen.getByLabelText(/radio label/i)).toBeInTheDocument()
    expect(screen.getByText(/radio label/i)).toHaveAttribute('for', 'radio')

    expect(container.firstChild).toMatchSnapshot()
  })

  it('should render without label', () => {
    renderWithTheme(<Radio />)

    expect(screen.queryByRole('label')).not.toBeInTheDocument()
  })

  it('should render with black label', () => {
    renderWithTheme(
      <Radio label="Radio label" labelFor="radio" labelColors="black" />
    )

    expect(screen.getByText(/radio label/i)).toHaveStyle({
      color: theme.colors.black
    })
  })

  it('should dispatch onCheck when label status changes', async () => {
    const onCheck = jest.fn()
    renderWithTheme(
      <Radio
        label="Radio label"
        labelFor="radio"
        onCheck={onCheck}
        value="anyValue"
      />
    )

    expect(onCheck).not.toHaveBeenCalled()

    userEvent.click(screen.getByLabelText('Radio label'))
    await waitFor(() => {
      expect(onCheck).toHaveBeenCalledTimes(1)
    })
    expect(onCheck).toHaveBeenCalledWith('anyValue')
  })

  it('should be accessible with tab', () => {
    renderWithTheme(<Radio label="Radio label" labelFor="radio" />)

    expect(document.body).toHaveFocus()

    userEvent.tab()

    expect(screen.getByLabelText(/radio label/i)).toHaveFocus()
  })
})
