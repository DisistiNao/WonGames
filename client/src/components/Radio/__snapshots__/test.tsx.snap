// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Radio /> should render with label 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c1 {
  cursor: pointer;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  width: 1.8rem;
  height: 1.8rem;
  border: 0.2rem solid #2E2F42;
  border-radius: 50%;
  outline: none;
}

.c1:focus {
  box-shadow: 0 0 0.5rem #F231A5;
}

.c1::after {
  background: #F231A5;
  content: '';
  position: absolute;
  width: 1rem;
  height: 1rem;
  border-radius: 50%;
  -webkit-transition: opacity 0.1s ease-in-out;
  transition: opacity 0.1s ease-in-out;
  opacity: 0;
}

.c1:checked {
  border-color: #F231A5;
  position: relative;
}

.c1:checked::after {
  opacity: 1;
}

.c2 {
  cursor: pointer;
  padding-left: 0.8rem;
  color: #FAFAFA;
  line-height: 1.8rem;
}

<main
  class="c0"
>
  <input
    class="c1"
    id="radio"
    type="radio"
    value=""
  />
  <label
    class="c2"
    for="radio"
  >
    Radio label
  </label>
</main>
`;
